### PR TITLE
Added closure to configure template engines

### PR DESCRIPTION
**Pull request by sunng87**
Currently there is no way to access template engine, for example, adding custom helper to handlebars.

This patch added a closure to `Template::fairing` which allows a closure to configure those engines.

### CODE REVIEW

**Comment by SergioBenitez** [comment]
Can you remove all of the style changes? We don't use 'rustfmt' since it doesn't quite work well enough yet.

I can review the PR once only functional changes are present.

**Comment by sunng87** [comment]
@SergioBenitez disabled my `format-on-save` and commit again.

File: contrib/src/templates/context.rs [hunk 1]
```diff
@@ -49,6 +49,10 @@ impl Context {
             Context { root, templates, engines }
         })
     }
+
+    pub fn engines_mut(&mut self) -> &mut Engines {
```
Comment (SergioBenitez) [inline contrib/src/templates/context.rs]: Remove this method.

File: contrib/src/templates/engine.rs [hunk 2]
```diff
@@ -69,4 +69,14 @@ impl Engines {
 
         None
     }
+
+    #[cfg(feature = "handlebars_templates")]
+    pub fn handlebars(&mut self) -> &mut Handlebars {
```
Comment (SergioBenitez) [inline contrib/src/templates/engine.rs]: Remove this method.

File: contrib/src/templates/engine.rs [hunk 3]
```diff
@@ -69,4 +69,14 @@ impl Engines {
...
+        &mut self.handlebars
+    }
+
+    #[cfg(feature = "tera_templates")]
+    pub fn tera(&mut self) -> &mut Tera {
```
Comment (SergioBenitez) [inline contrib/src/templates/engine.rs]: Remove this method.

File: contrib/src/templates/mod.rs [hunk 4]
```diff
@@ -158,9 +158,37 @@ impl Template {
...
+    ///     rocket::ignite()
+    ///         // ...
+    ///         .attach(Template::config_and_fairing(|ctxt| {
+    ///             // ctxt.engines_mut().handlebars().register_helper ...
+    ///         }))
+    ///         // ...
+    ///     # ;
+    /// }
+    /// ```
+    pub fn config_and_fairing<F>(f: F) -> impl Fairing
```
Comment (SergioBenitez) [inline contrib/src/templates/mod.rs]: Name this method `custom`.

File: contrib/src/templates/mod.rs [hunk 5]
```diff
@@ -158,9 +158,37 @@ impl Template {
...
+        where F: Fn(&mut Context) + Send + Sync + 'static
```
Comment (SergioBenitez) [inline contrib/src/templates/mod.rs]: Pass in an `&mut Engines` instead. Make the fields of the `Engines` struct public.

File: examples/handlebars_templates/src/main.rs [hunk 6]
```diff
@@ -46,6 +47,27 @@ fn rocket() -> rocket::Rocket {
...
 
+// when you want to register handlebars helpers
+#[allow(dead_code)]
+fn rocket_with_hbs_helper() -> rocket::Rocket {
+    rocket::ignite()
+        .mount("/", routes![index, get])
+        .attach(Template::config_and_fairing(|ctxt| {
+            ctxt.engines_mut()
+                .handlebars()
+                .register_helper("test",
```
Comment (SergioBenitez) [inline examples/handlebars_templates/src/main.rs]: We need a real test here. Create a helper than does something, and use it in one of the templates. Ensure that the returned value is what's expected.

File: examples/handlebars_templates/src/main.rs [hunk 7]
```diff
@@ -46,6 +47,27 @@ fn rocket() -> rocket::Rocket {
         .catch(errors![not_found])
 }
 
+// when you want to register handlebars helpers
+#[allow(dead_code)]
+fn rocket_with_hbs_helper() -> rocket::Rocket {
```
Comment (SergioBenitez) [inline examples/handlebars_templates/src/main.rs]: Don't create a function that's not used. Instead, perhaps you can name this function something like:

```rust
fn install_helper(handlebars: &mut Handlebars) { .. }
```

 and then call it from `rocket()`:

```rust
.attach(Template::custom(|engines| install_helper(&mut engines.handlebars)));
```

File: contrib/src/lib.rs [hunk 8]
```diff
@@ -74,3 +74,9 @@ mod uuid;
 
 #[cfg(feature = "uuid")]
 pub use uuid::{UUID, UuidParseError};
+
+#[cfg(feature = "handlebars_templates")]
+pub extern crate handlebars;
```
Comment (SergioBenitez) [inline contrib/src/lib.rs]: Move these `extern crate`s to the top of the file below the existing `extern crate`s.

File: contrib/src/templates/mod.rs (hunk 5 above)
Comment (SergioBenitez) [reply contrib/src/templates/mod.rs]: Also, why does `F` have to be `Send + Sync + 'static`?

Comment (sunng87) [reply contrib/src/templates/mod.rs]: I think it `Adhoc::attach` requested. It won't compile if F is not marked `Send + Sync + 'static`

File: examples/handlebars_templates/src/main.rs [hunk 9]
```diff
@@ -42,7 +43,21 @@ fn not_found(req: &Request) -> Template {
...
+                                     if let Some(p0) = h.param(0) {
+                                         rc.writer.write(p0.value()
+                                                         .render()
+                                                         .into_bytes()
+                                                         .as_ref())?;
+                                     }
+                                     Ok(())
+                                 }));
+
+        }))
```
Comment (sunng87) [inline examples/handlebars_templates/src/main.rs]: I just changed default `rocket()` to use `custom`

**Comment by SergioBenitez** [comment]
Please see the build failure.

**Comment by sunng87** [comment]
@SergioBenitez it was `Template::show` that has no helper config. Will a API break change on `Template::show` work for you? I'm going to add a closure argument to it. Of course we can a `Template::custom_and_show` but I think it's a little verbose for a test method. WDYT?

**Comment by SergioBenitez** [comment]
So sorry for the delay here! As we spoke about briefly in person, I think the API for `Template::show` needs to change a bit. The upside is that I think it'll actually be cleaner.

Here's the signature I'm imagining for `Template::show()`:

```rust
    pub fn show<S, C>(rocket: &Rocket, name: S, context: C) -> Option<String>
        where S: Into<Cow<'static, str>>, C: Serialize
    {
```

The big change is that `show()` now takes in a `Rocket` instance explicitly. This means that we can use the cached versions of templates (yay!) along with their configuration (yay!) and carrying around a root path is no longer required (yay!). An instance of `Rocket` should be available at all points necessary. In tests, it can be obtained via `client.rocket()`.

What do you think?

**Comment by SergioBenitez** [comment]
Just a heads up that #500 also seeks to implement this.

**Pull request closed without merging**